test/dns-test.js

```diff
@@ -0,0 +1,396 @@
+
+    // record to save in the authenticated tree
+    for (const [hstype, item] of Object.entries(json)) {
+      const type = types[item.type];
+      const resource = Resource.fromJSON(item.resource);
+
+      // args: size, height, network
+      const name = rules.grindName(5, 1, network);
+
+      it(`should update ${name}`, async () => {
+        await wallet.client.post(`/wallet/${wallet.id}/open`, {
+          name: name
+        });
+
+        await mineBlocks(treeInterval + 1, coinbase);
+
+        await wallet.client.post(`/wallet/${wallet.id}/bid`, {
+          name: name,
+          bid: 1000,
+          lockup: 2000
+        });
+
+        await mineBlocks(biddingPeriod + 1, coinbase);
+
+        await wallet.client.post(`/wallet/${wallet.id}/reveal`, {
+          name: name
+        });
+
+        await mineBlocks(revealPeriod + 1, coinbase);
+
+        await wallet.client.post(`/wallet/${wallet.id}/update`, {
+          name: name,
+          data: resource.toJSON()
+        });
+
+        // mine a block
+        await mineBlocks(1, coinbase);
+
+        const rinfo = await nclient.execute('getnameresource', [name]);
+
+        const json = resource.getJSON(name);
+        assert.deepEqual(rinfo, json);
+
+        const info = await nclient.execute('getnameinfo', [name]);
+        const data = Buffer.from(info.info.data, 'hex');
+
+        const returned = Resource.decode(data);
+        assert.deepEqual(returned.getJSON(name), json);
+
+        // this commits the tree state to disk
+        // updates served by the dns server
+        await mineBlocks(treeInterval, coinbase);
+      });
+
+      it(`should return authenticated ${hstype} record`, async () => {
+        // Certain types of queries require the
+        // name to be formatted with additional data
+        const query = buildQuery(name, resource, type);
+
+        const response = await nstub.lookup(query, type);
+        const res = response.toJSON();
+        const dns = resource.toDNS(util.fqdn(name), type).toJSON();
+
+        // query for the zone signing key
+        const dnskey = await nstub.lookup('.', types.DNSKEY);
+
+        // parse the zone signing key and
+        // key signing key out of the response
+        const {zsk, ksk} = getSigningKeys(dnskey);
+
+        assert(zsk instanceof wire.Record);
+        assert(ksk instanceof wire.Record);
+
+        // validate the signature on the ZSK
+        verifyDNSSEC(dnskey, ksk, types.DNSKEY, '.');
+        // validate the signature over the rrsets
+        verifyDNSSEC(response, zsk, type, query);
+
+        // NOTE: the signatures are not canonical
+        // when the native backend is being used
+        // because it uses OpenSSL which does not
+        // yet use RFC 6979, so nullify the
+        // signatures before comparing them
+
+        if (bcrypto.native === 2)
+          assert.deepEqual(nullSig(res), nullSig(dns));
+        else
+          assert.deepEqual(res.name, dns.name);
+      });
```

Comment:
shouldn't this be `assert.deepEqual(res, dns);`?

Reply:
I can do that but would need to set the edns field along with a few bits in the header. You can see here when the tests run, there are additional fields that are set by the DNS server itself that are not set when calling `resource.toDNS`. Trimmed for brevity. This would expand the test coverage to include those fields being used by the DNS server.

```
  DNS Servers
    Authoritative Resolver
      ✓ should return no answer for HNS name pre-update
      1) should return authenticated INET4 record

  2 passing (948ms)
  20 failing

  1) DNS Servers
       Authoritative Resolver
         should return authenticated INET4 record:

      AssertionError [ERR_ASSERTION]: [Object] deepStrictEqual [Object]

      Expected values to be strictly deep-equal:
      
      + actual - expected ... Lines skipped
      
        {
          aa: true,
      ...
          authority: [],
      +   cd: true,
      -   cd: false,
          code: 'NOERROR',
          code: 'NOERROR',
      +   edns: {
      +     code: 0,
      +     dnssec: true,
      +     enabled: true,
      +     options: [],
      +     size: 4096,
      +     version: 0
      ...
      -   edns: undefined,
      -   id: 0,
      -   opcode: 'QUERY',
      -   qr: false,
      -   question: [],
      -   ra: false,
      -   rd: false,
      ...
      
      assert.deepEqual(res, dns);
             ^

      at Context.it (/home/tynes/Projects/github.com/handshake-org/hsd/test/dns-test.js:170:18)

```